cmd/k8s-deployment-scaler/main.go

```diff
@@ -0,0 +1,227 @@
+	if err := encodeAndWriteJSON(w, map[string]string{"status": "OK"}); err != nil {
+		log.Printf("Error in healthCheck: %v", err)
+		http.Error(w, "Internal server error", http.StatusInternalServerError)
+	}
+}
+
+// handleReplicaCount handles the /replica-count endpoint for both GET and POST requests.
+func handleReplicaCount(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	namespace := r.URL.Query().Get("namespace")
+	deployment := r.URL.Query().Get("deployment")
+
+	switch r.Method {
+	case http.MethodGet:
+		handleGetReplicaCount(w, namespace, deployment)
+	case http.MethodPost:
+		handlePostReplicaCount(w, r, namespace, deployment)
+	default:
+		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
+	}
+}
+
+func handleGetReplicaCount(w http.ResponseWriter, namespace, deployment string) {
+	if namespace == "" && deployment == "" {
+		response := map[string]interface{}{
+			"replicaCounts": map[string]int{
+				"default/my-deployment":                3,
+				"another-namespace/another-deployment": 5,
+			},
+		}
+		if err := encodeAndWriteJSON(w, response); err != nil {
+			log.Printf("Error in handleGetReplicaCount: %v", err)
+			http.Error(w, "Internal server error", http.StatusInternalServerError)
+		}
+	} else if namespace != "" && deployment != "" {
+		response := map[string]interface{}{
+			"replicaCount": 3, // Placeholder response
+		}
+		if err := encodeAndWriteJSON(w, response); err != nil {
+			log.Printf("Error in handleGetReplicaCount: %v", err)
+			http.Error(w, "Internal server error", http.StatusInternalServerError)
+		}
+	} else {
+		http.Error(w, "Both namespace and deployment must be specified together", http.StatusBadRequest)
+	}
+}
+
+func handlePostReplicaCount(w http.ResponseWriter, r *http.Request, namespace, deployment string) {
+	if namespace == "" || deployment == "" {
+		http.Error(w, "Missing query parameters", http.StatusBadRequest)
+		return
+	}
+
+	var reqBody struct {
+		Replicas int `json:"replicas"`
+	}
+
+	if err := json.NewDecoder(r.Body).Decode(&reqBody); err != nil {
+		log.Printf("Error parsing request body: %v", err)
+		http.Error(w, "Invalid request body", http.StatusBadRequest)
+		return
+	}
+
+	if reqBody.Replicas < 0 {
+		http.Error(w, "Replica count must be non-negative", http.StatusBadRequest)
+		return
+	}
+
+	log.Printf("Setting replica count for %s/%s to %d", namespace, deployment, reqBody.Replicas)
+
+	response := map[string]interface{}{
+		"replicaCount": reqBody.Replicas,
+	}
+	if err := encodeAndWriteJSON(w, response); err != nil {
+		log.Printf("Error in handlePostReplicaCount: %v", err)
+		http.Error(w, "Internal server error", http.StatusInternalServerError)
+	}
+}
+
+// listDeployments handles the /deployments endpoint to list deployments
+func listDeployments(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
```

Comment:
While explicitly checking for HTTP request type in handler works just fine, it adds more boilerplate to the implementation. Any ideas about better way to handle this?

Reply:
I could use a middleware to achieve a DRY implementation while still checking the request method.

Reply:
You could just handle it on the `mux` level, something like:

```
mux.HandleFunc("GET /healthz", healthCheck)
```

See some examples here:
- https://eli.thegreenplace.net/2023/better-http-server-routing-in-go-122

Reply:
Thank you for pointing that out, I went with your recommendation.